src/main/java/gregtech/api/util/virtualregistry/VirtualRegistryMap.java

```diff
@@ -0,0 +1,83 @@
+package gregtech.api.util.virtualregistry;
+
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.common.util.INBTSerializable;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class VirtualRegistryMap implements INBTSerializable<NBTTagCompound> {
+
+    private final Map<EntryTypes<?>, Map<String, VirtualEntry>> registryMap = new HashMap<>();
+
+    public VirtualRegistryMap(NBTTagCompound tag) {
+        deserializeNBT(tag);
+    }
+
+    public VirtualRegistryMap() {}
+
+    @SuppressWarnings("unchecked")
+    public <T extends VirtualEntry> T getEntry(EntryTypes<T> type, String name) {
+        return (T) registryMap.get(type).get(name);
```

Comment:
Should we null check the first `get()`? Applies to a lot of places in this file.